### PR TITLE
NIFI-8616 - Migrate PutTCP, PutUDP and PutSplunk to use Netty

### PR DESCRIPTION
**Pull request by thenatog**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by thenatog** [comment]
Just doing some further testing on this but wanted to get a PR up for feedback.

File: nifi-commons/nifi-socket-utils/src/main/java/org/apache/nifi/io/socket/SocketUtils.java [hunk 1]
```diff
@@ -188,6 +188,9 @@ public static void closeQuietly(final Socket socket) {
                 }
             } finally {
                 if (!socket.isClosed()) {
+                    if (socket.getChannel().isOpen()) {
+                        socket.getChannel().close();
+                    }
```
Comment (exceptionfactory) [inline nifi-commons/nifi-socket-utils/src/main/java/org/apache/nifi/io/socket/SocketUtils.java:193]: Is this change necessary?  According to the `java.net.Socket` documentation, `getChannel()` can return null, so if this change is necessary, the result of `getChannel()` should be checked for `null`.

File: nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/ByteArrayNettyEventSenderFactory.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import io.netty.channel.ChannelHandler;
+import io.netty.handler.codec.bytes.ByteArrayEncoder;
+import org.apache.nifi.event.transport.configuration.TransportProtocol;
+import org.apache.nifi.event.transport.netty.channel.LogExceptionChannelHandler;
+import org.apache.nifi.logging.ComponentLog;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Netty Event Sender Factory for String messages
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/ByteArrayNettyEventSenderFactory.java:28]: The comment should be updated to reflect implementation:
```suggestion
 * Netty Event Sender Factory for byte array messages
```

File: nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/NettyEventSenderFactory.java [hunk 3]
```diff
@@ -49,6 +49,7 @@
  */
 public class NettyEventSenderFactory<T> extends EventLoopGroupFactory implements EventSenderFactory<T> {
     private static final int MAX_PENDING_ACQUIRES = 1024;
+    private final static int FRAME_SIZE = 40000;
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/NettyEventSenderFactory.java:52]: Is there a particular reason for this specific frame size?  It seems like it should be a configurable `Integer` property that could be checked for null and used to set the Send and Receive Buffer sizes when found.

File: nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/NettyEventSenderFactory.java [hunk 4]
```diff
@@ -116,6 +117,8 @@ public void setMaxConnections(final int maxConnections) {
         bootstrap.remoteAddress(new InetSocketAddress(address, port));
         final EventLoopGroup group = getEventLoopGroup();
         bootstrap.group(group);
+        bootstrap.option(ChannelOption.SO_SNDBUF, FRAME_SIZE);
+        bootstrap.option(ChannelOption.SO_RCVBUF, FRAME_SIZE);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/NettyEventSenderFactory.java:121]: Is this necessary for all uses of the Sender Factory?  Changing the value to a nullable Integer would make it more configurable if it can be optional.

File: nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/put/AbstractPutEventProcessor.java [hunk 5]
```diff
@@ -85,16 +86,15 @@
...
     public static final AllowableValue TCP_VALUE = new AllowableValue("TCP", "TCP");
     public static final AllowableValue UDP_VALUE = new AllowableValue("UDP", "UDP");
-
     public static final PropertyDescriptor PROTOCOL = new PropertyDescriptor
             .Builder().name("Protocol")
             .description("The protocol for communication.")
             .required(true)
             .allowableValues(TCP_VALUE, UDP_VALUE)
             .defaultValue(TCP_VALUE.getValue())
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/put/AbstractPutEventProcessor.java:97]: Is this necessary since allowable values are specified and the property is required?

File: nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/main/java/org/apache/nifi/processors/splunk/PutSplunk.java [hunk 6]
```diff
@@ -136,45 +117,38 @@ public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFacto
...
         }
 
-        // get a sender from the pool, or create a new one if the pool is empty
-        // if we can't create a new connection then route flow files to failure and yield
-        // acquireSender will handle the routing to failure and yielding
-        ChannelSender sender = acquireSender(context, session, flowFile);
-        if (sender == null) {
+        if (eventSender == null) {
             return;
         }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/main/java/org/apache/nifi/processors/splunk/PutSplunk.java:127]: Is it possible for `eventSender` to be `null` here, or can this check be removed?

File: nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/main/java/org/apache/nifi/processors/splunk/PutSplunk.java [hunk 7]
```diff
@@ -136,45 +117,38 @@ public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFacto
...
+            session.commitAsync();
+            context.yield();
         }
     }
 
     /**
      * Send the entire FlowFile as a single message.
      */
-    private void processSingleMessage(ProcessContext context, ProcessSession session, FlowFile flowFile, ChannelSender sender) {
+    private void processSingleMessage(final ProcessContext context, final ProcessSession session, final FlowFile flowFile, final EventSender<byte[]> sender) {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/main/java/org/apache/nifi/processors/splunk/PutSplunk.java:151]: Is it necessary to pass `EventSender` as parameter here since it is already a class member?

File: nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/main/java/org/apache/nifi/processors/splunk/PutSplunk.java [hunk 8]
```diff
@@ -200,20 +174,15 @@ public void process(final InputStream in) throws IOException {
...
+        sender.sendEvent(buf);
+        messageBatch.addSuccessfulRange(0L, flowFile.getSize());
     }
 
     /**
      * Read delimited messages from the FlowFile tracking which messages are sent successfully.
      */
     private void processDelimitedMessages(final ProcessContext context, final ProcessSession session, final FlowFile flowFile,
-                                          final ChannelSender sender, final String delimiter) {
+                                          final EventSender<byte[]> sender, final String delimiter) {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/main/java/org/apache/nifi/processors/splunk/PutSplunk.java:185]: Is it necessary to pass `EventSender` as parameter here since it is already a class member?

File: nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/test/java/org/apache/nifi/processors/splunk/TestPutSplunk.java [hunk 9]
```diff
@@ -317,84 +260,39 @@ public void testUnableToCreateConnectionShouldRouteToFailure() {
...
-    /**
-     * Extend PutSplunk to use a CapturingChannelSender.
-     */
-    private static class UnableToConnectPutSplunk extends PutSplunk {
-
-        @Override
-        protected ChannelSender createSender(String protocol, String host, int port, int timeout, int maxSendBufferSize, SSLContext sslContext) throws IOException {
-            throw new IOException("Unable to create connection");
-        }
+    private void createTestServer(final String address, final TransportProtocol protocol) {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/test/java/org/apache/nifi/processors/splunk/TestPutSplunk.java:263]: Is `address` always `LOCALHOST`?  If so, then the `address` parameter could be removed in this method and the other `createTestServer` methods.

File: nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/test/java/org/apache/nifi/processors/splunk/TestPutSplunk.java [hunk 10]
```diff
@@ -317,84 +260,39 @@ public void testUnableToCreateConnectionShouldRouteToFailure() {
...
+        NettyEventServerFactory serverFactory = new ByteArrayMessageNettyEventServerFactory(runner.getLogger(), address, port, protocol, delimiter, VALID_LARGE_FILE_SIZE, messages);
+        if (sslContext != null) {
+            serverFactory.setSslContext(sslContext);
         }
+        eventServer = serverFactory.getEventServer();
+    }
 
-        public void setErrorStart(int errorStart) {
-            this.errorStart = errorStart;
+    private void checkReceivedAllData(final String[] sentData) throws Exception {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/test/java/org/apache/nifi/processors/splunk/TestPutSplunk.java:287]: The `String[]` parameter could be changed to use `String... sentData` variable arguments, which would remove the need for calling methods to pass a `String[]`.

File: nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/test/java/org/apache/nifi/processors/splunk/TestPutSplunk.java [hunk 11]
```diff
@@ -317,84 +260,39 @@ public void testUnableToCreateConnectionShouldRouteToFailure() {
...
+            ByteArrayMessage packet = messages.take();
+            assertNotNull(packet);
+            assertArrayEquals(item.getBytes(), packet.getMessage());
         }
 
-        public void setErrorEnd(int errorEnd) {
-            this.errorEnd = errorEnd;
-        }
+        // Check that we have no unexpected extra data.
+        assertNull(messages.poll());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-splunk-bundle/nifi-splunk-processors/src/test/java/org/apache/nifi/processors/splunk/TestPutSplunk.java:296]: It might be helpful to add a message here, then the comment could be removed:
```suggestion
        assertNull("Unprocessed messages found", messages.poll());
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutTCP.java [hunk 12]
```diff
@@ -168,61 +144,23 @@ public void onTrigger(final ProcessContext context, final ProcessSessionFactory
...
-                context.yield();
-            }
             return;
         }
 
-        ChannelSender sender = acquireSender(context, session, flowFile);
-        if (sender == null) {
+        if (eventSender == null) {
             return;
         }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutTCP.java:152]: Is it possible for eventSender to be null, or can this be removed?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutTCP.java [hunk 13]
```diff
@@ -168,61 +144,23 @@ public void onTrigger(final ProcessContext context, final ProcessSessionFactory
...
-                    final Charset charSet = Charset.forName(context.getProperty(CHARSET).getValue());
-                    out.write(delimiter.getBytes(charSet), 0, delimiter.length());
-                }
-                out.flush();
-            } catch (final Exception e) {
-                closeSender = true;
-                throw e;
-            }
-
+            byte[] content = readContent(context, session, flowFile);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutTCP.java:155]: Recommend adding `final`:
```suggestion
            final byte[] content = readContent(context, session, flowFile);
```

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutTCP.java [hunk 14]
```diff
@@ -168,61 +144,23 @@ public void onTrigger(final ProcessContext context, final ProcessSessionFactory
...
+            StopWatch stopWatch = new StopWatch(true);
+            eventSender.sendEvent(content);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutTCP.java:157]: Although this fits the pattern of sending individual messages as `byte[]` objects, this changes the behavior of the processor to read the entire FlowFile into memory.  Changing the behavior to read the FlowFile input stream in `byte[]` chunks may be one solution, but it may be necessary to have a different Netty-based handler for streams.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutTCP.java [hunk 15]
```diff
@@ -96,29 +92,9 @@
...
-        final SSLContextService sslContextService = (SSLContextService) context.getProperty(SSL_CONTEXT_SERVICE).asControllerService();
-
-        SSLContext sslContext = null;
-        if (sslContextService != null) {
-            sslContext = sslContextService.createContext();
-        }
-
-        return createSender(protocol, hostname, port, timeout, bufferSize, sslContext);
+    protected void init(final ProcessorInitializationContext context) {
+        super.init(context);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutTCP.java:97]: Is this method necessary if it is just calling super.init()?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutUDP.java [hunk 16]
```diff
@@ -81,22 +80,9 @@
...
     @Override
-    protected ChannelSender createSender(final ProcessContext context) throws IOException {
-        final String protocol = UDP_VALUE.getValue();
-        final String hostname = context.getProperty(HOSTNAME).evaluateAttributeExpressions().getValue();
-        final int port = context.getProperty(PORT).evaluateAttributeExpressions().asInteger();
-        final int bufferSize = context.getProperty(MAX_SOCKET_SEND_BUFFER_SIZE).asDataSize(DataUnit.B).intValue();
-
-        return createSender(protocol, hostname, port, 0, bufferSize, null);
+    protected void init(final ProcessorInitializationContext context) {
+        super.init(context);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutUDP.java:85]: Is this method necessary if it is just calling super.init()?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutUDP.java [hunk 17]
```diff
@@ -131,31 +117,25 @@ public void onTrigger(final ProcessContext context, final ProcessSessionFactory
...
-                context.yield();
-            }
             return;
         }
 
-        ChannelSender sender = acquireSender(context, session, flowFile);
-        if (sender == null) {
+        if (eventSender == null) {
             return;
         }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutUDP.java:125]: As mentioned on other classes, not sure if this check is necessary.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutTCP.java [hunk 18]
```diff
@@ -189,67 +181,64 @@ public void testRunSuccessConnectionFailure() throws Exception {
...
 
-    private void createTestServer(final String delimiter) throws Exception {
-        createTestServer(delimiter, false);
-    }
-
-    private void createTestServer(final String delimiter, final boolean closeOnMessageReceived) throws Exception {
-        createTestServer(delimiter, closeOnMessageReceived, ServerSocketFactory.getDefault());
+    private void createTestServer(final String address, final int port, final SSLContext sslContext) throws Exception {
+        messages = new LinkedBlockingQueue<>();
+        final byte[] delimiter = getDelimiter();// OUTGOING_MESSAGE_DELIMITER.getBytes(CHARSET);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutTCP.java:220]: Is the a reason for the associated comment?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutUDP.java [hunk 19]
```diff
@@ -120,29 +125,25 @@ public void testEmptyFile() throws Exception {
...
     }
 
+    @Ignore("This test is timing out but should pass. Needs fixing.")
     @Test(timeout = DEFAULT_TEST_TIMEOUT_PERIOD)
-    public void testlargeValidFile() throws Exception {
+    public void testLargeValidFile() throws Exception {
         configureProperties(UDP_SERVER_ADDRESS, true);
         final String[] testData = createContent(VALID_LARGE_FILE_SIZE);
         sendTestData(testData);
+        //checkRelationships(1, testData.length);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutUDP.java:134]: Is this commented line going to be removed, or is there an issue with the test?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutUDP.java [hunk 20]
```diff
@@ -190,15 +191,17 @@ public void testLoadTest() throws Exception {
...
         removeTestServer();
-        createTestServer(address, port, recvQueueSize);
+        createTestServer(address, port, frameSize);
     }
 
     private void configureProperties(final String host, final boolean expectValid) {
         runner.setProperty(PutUDP.HOSTNAME, host);
-        runner.setProperty(PutUDP.PORT, Integer.toString(server.getLocalPort()));
+        runner.setProperty(PutUDP.PORT, Integer.toString(port));
+        runner.setProperty(PutUDP.MAX_SOCKET_SEND_BUFFER_SIZE, "5 MB");
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/TestPutUDP.java:203]: Is there a reason for setting `5 MB` here?  That seems larger than theoretically possible given the maximum size of UDP packets.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for putting this together @thenatog, it definitely cleans up these components!  I noted a handful of details, and one particular question regarding stream handling for `PutTCP`.

File: nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/StreamingNettyEventSenderFactory.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+     * @param port Remote Port Number
+     * @param protocol Channel Protocol
+     */
+    public StreamingNettyEventSenderFactory(final ComponentLog log, final String address, final int port, final TransportProtocol protocol) {
+        super(address, port, protocol);
+        final List<ChannelHandler> handlers = new ArrayList<>();
+        handlers.add(new LogExceptionChannelHandler(log));
+        handlers.add(new ChunkedWriteHandler());
+        handlers.add(new InputStreamMessageEncoder());
+        setHandlerSupplier(() -> handlers);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/StreamingNettyEventSenderFactory.java:48]: Netty indicates that `ChunkedWriteHandler` is not a `@Sharable` handler, so it needs to be instantiated inside the handler supplier callback method in order for it to be created when needed.  `InputStreamMessageEncoder` can be annotated with `@ChannelHandler.Sharable` since it is sharable across channels.
```suggestion
        final LogExceptionChannelHandler logExceptionChannelHandler = new LogExceptionChannelHandler(log);
        final InputStreamMessageEncoder inputStreamMessageEncoder = new InputStreamMessageEncoder();
        setHandlerSupplier(() -> Arrays.asList(
            logExceptionChannelHandler,
            new ChunkedWriteHandler(),
            inputStreamMessageEncoder
        ));
```

File: nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/codec/InputStreamMessageEncoder.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.nifi.event.transport.netty.DelimitedInputStream;
+
+import java.io.InputStream;
+import java.util.List;
+
+/**
+ * Message encoder for an InputStream, which wraps the stream in a ChunkedStream for use with a ChunkedWriter. Can add a delimiter
+ * to the end of the output objects if the InputStream is a DelimitedInputStream.
+ */
+public class InputStreamMessageEncoder extends MessageToMessageEncoder<InputStream> {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-extension-utils/nifi-event-transport/src/main/java/org/apache/nifi/event/transport/netty/codec/InputStreamMessageEncoder.java:32]: This class can be annotated with `@ChannelHandler.Sharable` so that Netty can reuse the same instance.
```suggestion
@ChannelHandler.Sharable
public class InputStreamMessageEncoder extends MessageToMessageEncoder<InputStream> {
```

**Comment by exceptionfactory** [comment]
Thanks for the adjustments and updates @thenatog! Verified PutUDP and PutTCP with various settings. +1 Merging.

**Review by exceptionfactory — APPROVED** [review]
